### PR TITLE
Implement Geo types

### PR DESCRIPTION
**Pull request by guilhermeleobas**
This is a follow-up PR of #523 which adds support for all the geo types implemented in HeavyDB. To execute, it requires PR https://github.com/heavyai/heavydb-internal/pull/7126.

* GeoPoint
* GeoMultiPoint
* GeoLineString
* GeoMultiLineString
* GeoPolygon
* GeoMultiPolygon

Rendered docs gives the list of implemented methods and attributes.

### CODE REVIEW

File: rbc/heavydb/geo_nested_array.py
Comment (brenocfg) [inline rbc/heavydb/geo_nested_array.py:111]: I'm assuming you tried passing the indices as an array first, then decided to change to 4 separate ints? If so I guess this can be removed?

**Review by brenocfg — APPROVED** [review]
Most of the Python code admittedly went a little over my head, so I'll let @pearu review those bits 😅.

The LLVM/Numba codegen parts look good to me though! Only had a couple of very small nits. Awesome stuff @guilhermeleobas.

File: rbc/heavydb/geo_nested_array.py
Comment (pearu) [inline rbc/heavydb/geo_nested_array.py:107]: Nit:
```suggestion
        index_t = typesystem.Type.fromstring("int64_t index_")
```

File: rbc/heavydb/geo_nested_array.py
Comment (pearu) [reply rbc/heavydb/geo_nested_array.py:111]: It is not clear to me what can be removed.

I guess this uses an assumption that llvm (or smth) will interpret the struct `{..., int64_t index[4], ...}` as `{..., int64_t index0, int64_t index1, int64_t index2, int64_t index4, ...}`.

File: rbc/heavydb/geo_nested_array.py
Comment (pearu) [inline rbc/heavydb/geo_nested_array.py:129]: why `double`? IIUC, it should be related to the bitwidth of `typ`.

File: rbc/heavydb/geo_nested_array.py
Comment (pearu) [inline rbc/heavydb/geo_nested_array.py:130]: Not likely. IIUC, the allocated memory should be registered for free, similar to what we have in `TextEncodingNone` support.

File: rbc/heavydb/geo_nested_array.py
Comment (pearu) [inline rbc/heavydb/geo_nested_array.py:232]: Just a note: `n_matrices`, `nrows`, and `ncols` are confusing variable names for nested arrays that can have variable nof "columns" and that dimensionality may be greater than 2, imho.
A better set of names would be `size0`, `size1`, etc, or `dim0`, `dim1`, etc, or similar.

File: rbc/heavydb/geo_nested_array.py
Comment (pearu) [inline rbc/heavydb/geo_nested_array.py:289]: General note: these loops are run in Python. Is this true or is the `impl` function jitted with numba?
If true, this implies that using `fromCoords` will not be the most efficient way to construct Geo data instances from Python.

File: rbc/heavydb/geo_nested_array.py
Comment (pearu) [inline rbc/heavydb/geo_nested_array.py:336]: Could you create an rbc issue about this?

File: rbc/heavydb/remoteheavydb.py
Comment (pearu) [inline rbc/heavydb/remoteheavydb.py:1454]: Why this change?
This will contradict the comment above as the action here is "convert `T foo` to `T foo | name=foo`".

File: rbc/heavydb/table_function_manager.py
Comment (pearu) [inline rbc/heavydb/table_function_manager.py:73]:
```suggestion
        values in the index-th output Column of varlen types.
```

File: rbc/libfuncs.py
Comment (pearu) [inline rbc/libfuncs.py:59]: A note: this list of function names depends on the heavydb version. I wonder if this fact is practically relevant.

File: rbc/tests/__init__.py
Comment (pearu) [inline rbc/tests/__init__.py:443]: Add also the corresponding doc lines for `multilinestring` and `multipolygon` for completeness.

**Review by pearu — CHANGES_REQUESTED** [review]
Thanks, @guilhermeleobas! Nice work!

I have a few nits but overall looks good.

**Comment by pearu** [comment]
When running rbc tests agains cuda-enabled heavydb server (master), the following tests fail:

<details>

```
FAILED rbc/tests/heavydb/test_geolinestring.py::test_ct_pointn[multipoint-mp2-1] - AssertionError: 
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict[8] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available f...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict[16] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available ...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict[32] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available ...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict2[8] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available ...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict2[16] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict2[32] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict3[8-base] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Avail...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict3[8-other] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Avai...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict3[8-another] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Av...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict3[16-base] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Avai...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict3[16-other] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Ava...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict3[16-another] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. A...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict3[32-base] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Avai...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict3[32-other] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Ava...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_shared_dict3[32-another] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. A...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_column_list[1-8] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_column_list[1-16] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Availabl...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_column_list[1-32] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Availabl...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_column_list[2-8] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_column_list[2-16] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Availabl...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_column_list[2-32] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Availabl...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_column_list[3-8] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_column_list[3-16] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Availabl...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_column_list[3-32] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Availabl...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_column_list2[32] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_count[8] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available functio...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_count[16] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available functi...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_count[32] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available functi...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_order_by[8] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available func...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_order_by[16] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available fun...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_text_encoding_order_by[32] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available fun...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId[t1-foo] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function t...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId[t1-bar] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function t...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId[t1-invalid1234] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available fu...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId[t2-foo] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function t...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId[t2-bar] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function t...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId[t2-invalid1234] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available fu...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId[t4-foo] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function t...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId[t4-bar] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function t...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId[t4-invalid1234] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available fu...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId_unicode[t1] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available functi...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId_unicode[t2] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available functi...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getStringId_unicode[t4] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available functi...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_empty_input_id - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function types:
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getString[t1] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function types:
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getString[t2] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function types:
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getString[t4] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function types:
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getString_lst[t1] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function types:
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getString_lst[t2] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function types:
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_getString_lst[t4] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Available function types:
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_udf_copy_dict_encoded_string[t1] - TypeError: found no matching function type to given argument types `int8*, int32` and device `GPU`. Availab...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_to_text_encoding_none[to_text_encoding_none_1-t1] - TypeError: found no matching function type to given argument types `int8*, int32` and devi...
FAILED rbc/tests/heavydb/test_text_encoding_dict.py::test_to_text_encoding_none[to_text_encoding_none_2-t1] - TypeError: found no matching function type to given argument types `int8*, int32` and devi...
FAILED rbc/tests/heavydb/test_udtf.py::test_table_function_manager - TypeError: found no matching function type to given argument types `int8*, int64` and device `GPU`. Available function types:
=========================================================== 55 failed, 1263 passed, 176 skipped, 6 xfailed, 466 warnings in 2041.04s (0:34:01) ============================================================
```

</details>

File: rbc/heavydb/geo_nested_array.py
Comment (guilhermeleobas) [reply rbc/heavydb/geo_nested_array.py:129]: You're right, it should be `i64(sizeof(typ_ptr))`, since we are creating a pointer here.

File: rbc/heavydb/geo_nested_array.py
Comment (guilhermeleobas) [reply rbc/heavydb/geo_nested_array.py:130]: So, I checked and all calls to `allocate_varlen_buffer` are registered to be freed by a pass in HeavyDB.

File: rbc/heavydb/geo_nested_array.py
Comment (guilhermeleobas) [reply rbc/heavydb/geo_nested_array.py:289]: The function is jitted with Numba

File: rbc/heavydb/column_flatbuffer.py
Comment (guilhermeleobas) [reply rbc/heavydb/column_flatbuffer.py:154]: Yes! The "problematic" SROA pass runs on -O2 as well.

File: rbc/heavydb/geo_nested_array.py
Comment (guilhermeleobas) [reply rbc/heavydb/geo_nested_array.py:111]: I think Breno mentioned the `# int64_t index[4]` line

> I'm assuming you tried passing the indices as an array first, then decided to change to 4 separate ints? If so I guess this can be removed?

Exactly

File: rbc/heavydb/remoteheavydb.py
Comment (guilhermeleobas) [reply rbc/heavydb/remoteheavydb.py:1454]: Reverted this change

File: rbc/heavydb/geo_nested_array.py
Comment (guilhermeleobas) [reply rbc/heavydb/geo_nested_array.py:336]: Created #543

File: rbc/typesystem.py
Comment (pearu) [inline rbc/typesystem.py:1078]: Nit:
```suggestion
        if s in cls.custom_types:
            return cls.custom_types[s](())
        a = cls.aliases.get(s)
        if a in cls.custom_types:
            return cls.custom_types[a](())
```

File: rbc/heavydb/geo_nested_array.py
Comment (pearu) [reply rbc/heavydb/geo_nested_array.py]: Great! Could you comment on this in rbc/heavydb/allocator.py? Something like:
```
When an LLVM module defines manage_memory_buffer=1, the memory buffers allocated by
allocate_varlen_buffer will be freed at the execution clean-up stage.
```

**Review by pearu — APPROVED** [review]
I have a couple of minor nits but LGTM! Thanks, @guilhermeleobas!

**Pull request merged**
